Dublin Core serializer: Transform identifiers to URLs if possible

max-moser opened the pull request:
Closes https://github.com/inveniosoftware/invenio-app-rdm/issues/1976

Adding prefixes to the `oai_dc` metadata format for OAI-PMH is a requirement we got from [BASE](https://www.base-search.net/).
This PR only touches the Dublin Core serializer in order to be minimally invasive.
It transforms the identifiers by trying to:
* Turn the identifier into a URL if possible (e.g. for DOIs and handles)
* Prefix the identifier with its scheme (e.g. the EAN-13 value `9780521425575` will become `ean13:9780521425575`) if that's not already the case (e.g. `ark:/123/456` will stay the same)

max-moser commented on invenio_rdm_records/resources/serializers/dublincore/schema.py:
```diff
@@ -56,11 +77,11 @@ def get_relations(self, obj):
 
         # Alternate identifiers
         for a in obj["metadata"].get("alternate_identifiers", []):
-            rels.append(f"{a['scheme']}:{a['identifier']}")
```
The old logic here would unconditionally append the scheme as a prefix, which looks weird in some cases.
For example, `ark:/123/456` would be turned into `ark:ark:/123/456` which seems redundant to me.

An EAN-13 value of e.g. `9780521425575` would be turned into `ean13:9780521425575` both before and after the PR.

ppanero commented on invenio_rdm_records/resources/serializers/dublincore/schema.py:
```diff
@@ -39,11 +40,31 @@ class DublinCoreSchema(BaseSchema):
-        items.extend(i["identifier"] for i in obj["metadata"].get("identifiers", []))
-        items.extend(p["identifier"] for p in obj.get("pids", {}).values())
+
+        for id_ in obj["metadata"].get("identifiers", []):
```
Does the order have any impact on the output? In general we try to give priority to `pids` since those fields have much more validation and logic rather than a simple format validation.  However, if the consumer of dc data does not have any implications due to order, then it is all the same. Otherwise, I'd inver the order (first pids, then identifiers).

I know it was potentially wrong before, but we can profit not and fix it hehe

max-moser replied:
sure thing, done :smile:

max-moser commented:
Updated the PR to control the behavior of `schema._transform_identifier()` via the `schema.context`.
Added a possibility to set the schema's context via the configuration, like:
```python
OAISERVER_METADATA_FORMATS = {
    "oai_dc": {
        "serializer": (
            "invenio_rdm_records.oai:dublincore_etree",
            {
                "schema_context": {
                    "prefix_identifier_schemes": True,
                    "urlize_identifiers": True,
                }
            },
        ),
        "schema": "http://www.openarchives.org/OAI/2.0/oai_dc.xsd",
        "namespace": "http://www.openarchives.org/OAI/2.0/oai_dc/",
    },
}
```
The relevant part here is that the value for `serializer` is a tuple with two entries.
The second entry is used as keyword argument for the function referenced in the first entry.

max-moser commented:
Ping @alejandromumo 
I couldn't quickly identify code changes related to https://github.com/inveniosoftware/invenio-rdm-records/issues/1231 but saw that you're working on it; could you please check if any changes here (e.g. building & passing the `schema_context`) interfere with that task?

max-moser commented on invenio_rdm_records/oai.py:
```diff
@@ -26,9 +26,9 @@
 
 
-def dublincore_etree(pid, record):
+def dublincore_etree(pid, record, **serializer_kwargs):
```
In case the `serializer` value for an OAI metadata format in the `OAISERVER_METADATA_FORMATS` config item happens to be a tuple, the second part of that tuple is being passed as a keyword argument to the function referenced by the first part. See: https://github.com/inveniosoftware/invenio-oaiserver/blob/master/invenio_oaiserver/utils.py#L53-L54

max-moser commented on invenio_rdm_records/oai.py:
```diff
@@ -26,9 +26,9 @@
+def dublincore_etree(pid, record, **serializer_kwargs):
     """Get DublinCore XML etree for OAI-PMH."""
-    json = DublinCoreXMLSerializer().dump_obj(record["_source"])
+    json = DublinCoreXMLSerializer(**serializer_kwargs).dump_obj(record["_source"])
```
So if the second part of the tuple has the shape `{"schema_context": {...}}`, it will be passed to the schema as `context` here via the serializer (see the added passing of `**options` in `invenio_rdm_records/resources/serializers/dublincore/__init__.py` below).

alejandromumo commented:
> Ping @alejandromumo I couldn't quickly identify code changes related to #1231 but saw that you're working on it; could you please check if any changes here (e.g. building & passing the `schema_context`) interfere with that task?

#1231 is still a WIP, but I don't think the changes you implemented will interfere with that.

max-moser commented:
ping @slint thoughts & prayers?

slint reviewed (approved):
LGTM, and pretty sane for customizing
